java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+package arcs.core.testutil
+
+import java.time.LocalDateTime
+import kotlin.random.Random
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.runBlocking
+
+/**
+ * This file contains utility classes and methods that establish fuzz-testing infrastructure.
+ *
+ * To write fuzz tests for a unit or assembly:
+ * (1) think of the invariants that make sense for the thing that you want to test.
+ *
+ * For example, if you've built a bank account class, then an invariant might be that the balance
+ * can never be less than 0. (See below for more information on invariants).
+ *
+ * (2) express this invariant as a function, ensuring all non-required structure is externalized.
+ *
+ * An invariant is a statement that's expected to be true, regardless of some context. The idea
+ * here is to make sure that the context (as much as possible) is passed into the invariant
+ * function, rather than choosing a specific context to test.
+ *
+ * For our bank account class, we want to run a sequence of transactions against an account and
+ * then test that the balance isn't negative. The sequence of transactions and the initial balance
+ * are both structure that is incidental to the invariant, so we will make these function
+ * parameters:
+ *
+ * fun invariant_balanceCantGoNegative(initialBalance: Int, transactions: List<Transaction>) {
+ *   val account = Account(initialBalance)
+ *   val transactionsApplied = transactions.map { account.apply(it) }
+ *   assertThat(account.balance).isLargerThan(0)
+ * }
+ *
+ * Generally, the invariant will separate out into three pieces: firstly there may be some
+ * preconditions that need to be checked (for example, we can't guarantee that the balance
+ * will be larger than 0 if the initialBalance is not). Then there will be the body of the
+ * invariant (applying the arguments, running the test, etc.); and finally there will be a
+ * set of postconditions which need to be asserted.
+ *
+ * Splitting the above function up gives us:
+ *
+ * fun invariant_balanceCantGoNegative(initialBalance: Int, transactions: List<Transaction>) {
+ *   // PRECONDITION
+ *   assertThat(initialBalance).isLargerThan(0)
+ *
+ *   // BODY
+ *   val account = Account(initialBalance)
+ *   val transactionsApplied = transactions.map { account.apply(it) }
+ *
+ *   // POSTCONDITION
+ *   assertThat(account.balance).isLargerThan(0)
+ * }
+ *
+ * In order to render these invariants useful, we also need to make sure that we can pass in
+ * random *generators* of context, rather than specific values. The interface A<Type> gives us a
+ * way of doing this; instances of Type can be recovered from A<Type> by invoking it. This
+ * sometimes introduces the need for an additional section in our invariant function that does
+ * setup:
+ *
+ * fun invariant_balanceCantGoNegative(initialBalance: A<Int>, transactions: A<List<Transaction>>)
```

Comment:
Reading this file from top to bottom, when I get here I don't know what A is. Should this be Generator<Int> or something like this? also it is not clear whether it is ok to call it multiple times or not. Would it be simple to pass the values directly, eg
fun invariant_balanceCantGoNegative(initialBalance: Int, transactions: List<Transaction>) {}
@Test
 fuzz_balanceCantGoNegative = runFuzzTest {
   invariant_balanceCantGoNegative(generateNonNegativeInt(it), generateTransactions(it))
 }
I think this would still satisfy your A,B,C,D below.

Reply:
I like this idea a lot (in fact, I suggested a very similar thing in my comment below, with a slightly different structure).

For basic tests, you can inline them so it becomes just a single function:

```
@Test
fun fuzz_balanceCantGoNegative = runFuzzTest { generator ->
  val initialBalance = generator.generateNonNegativeInt()
  val transactions = generator.generateTransactions()
  assertThat(initialBalance).isLargerThan(0)
  
  val account = Account(initialBalance)
  val transactionsApplied = transactions.map { account.apply(it) }

  assertThat(account.balance).isLargerThan(0)
}
```

This reads exactly like a regular unit test, except the inputs are obviously generated.

With some kotlin lambda trickery, you could hide the `generator ->` parameter and make it the `this` parameter, and then elide the `this`, to become: `val initialBalance = generateNonNegativeInt()`.

If you end up wanting to repeat the test for additional types of input data (e.g. the regression test described below), you can factor out the inner test code into its own function and call it from both places.

Reply:
I'm mostly OK with this suggestion (I think, I need to try it in code to see for sure).

Couple of points though
(1) there isn't just one generator, so you can't pass a generator into runFuzzTest. Each type has its own generator(s).
(2) generators are composable which may not fit well with them being functions. This is one of the things I need to play around with; for example does generateHandle(generateCreatableStorageKey(generateString()), generateType()) read OK?
(3) generators are (and should be) a source of values, not one-shot functions; they do have state. Otherwise generating random lists of like values gets quite difficult. This also makes them feel less like functions to me.
(4) Identifying invariants is a super important part of this process, to the extent that I strongly prefer them to always exist separate from the fuzz test body.